src/core/sensor/qgsabstractsensor.h

```diff
@@ -0,0 +1,208 @@
+    email                : mathieu at opengis dot ch
+ ***************************************************************************/
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSABSTRACTSENSOR_H
+#define QGSABSTRACTSENSOR_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgis.h"
+
+#include <QDateTime>
+#include <QDomElement>
+#include <QVariant>
+
+/**
+ * \ingroup core
+ * \class QgsAbstractSensor
+ * \brief An abstract base class for sensor classes
+ * \since QGIS 3.32
+ */
+class CORE_EXPORT QgsAbstractSensor : public QObject
+{
+#ifdef SIP_RUN
+#include "qgsiodevicesensor.h"
+#include "qgssensorregistry.h"
+#endif
+
+
+#ifdef SIP_RUN
+    SIP_CONVERT_TO_SUBCLASS_CODE
+    if ( QgsAbstractSensor *item = qobject_cast< QgsAbstractSensor * >( sipCpp ) )
+    {
+      if ( item->type() == QLatin1String( "tcp_socket" ) && dynamic_cast<QgsTcpSocketSensor *>( item ) != NULL )
+      {
+        sipType = sipType_QgsTcpSocketSensor;
+      }
+      else if ( item->type() == QLatin1String( "udp_socket" ) && dynamic_cast<QgsUdpSocketSensor *>( item ) != NULL )
+      {
+        sipType = sipType_QgsUdpSocketSensor;
+      }
+      else if ( item->type() == QLatin1String( "serial_port" ) && dynamic_cast<QgsSerialPortSensor *>( item ) != NULL )
+      {
+        sipType = sipType_QgsSerialPortSensor;
+      }
+      else
+      {
+        sipType = sipType_QgsAbstractSensor;
+      }
+    }
+    else
+    {
+      sipType = NULL;
+    }
+    SIP_END
+#endif
+
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Contains details of a sensor data capture
+     */
+    struct CORE_EXPORT SensorData
+    {
+
+      /**
+       * Last captured sensor value stored as a QVariant.
+       * \note The member can store multiple values if the sensor passes on a QVariantMap.
+       */
+      QVariant lastValue;
+
+      //! Timestamp of last captured sensor value
+      QDateTime lastTimestamp;
+    };
+
+    /**
+     * Constructor for an abstract sensor, bound to the specified \a parent.
+     */
+    explicit QgsAbstractSensor( QObject *parent SIP_TRANSFERTHIS = nullptr );
+    virtual ~QgsAbstractSensor() = default;
+
+    /**
+     * Returns the sensor ID.
+     */
+    QString id() const { return mId; }
+
+    /**
+     * Returns the sensor type
+     */
+    virtual QString type() const { return QString(); }
+
+    /**
+     * Returns the sensor name
```

Comment:
"This is a user-friendly string identifying the sensor, set by users when initially creating the sensor"